rllib/evaluation/worker_set.py

```diff
@@ -139,17 +139,34 @@ def remote_workers(self) -> List[ActorHandle]:
         """Returns a list of remote rollout workers."""
         return self._remote_workers
 
-    def sync_weights(self, policies: Optional[List[PolicyID]] = None) -> None:
+    def sync_weights(self,
+                     policies: Optional[List[PolicyID]] = None,
+                     from_worker: Optional[RolloutWorker] = None) -> None:
         """Syncs model weights from the local worker to all remote workers.
 
         Args:
-            policies: An optional list of policy IDs to sync for. If None,
-                sync all policies.
+            policies: Optional list of PolicyIDs to sync weights for.
+                If None (default), sync weights to/from all policies.
+            from_worker: Optional RolloutWorker instance to sync from.
+                If None (default), sync from this WorkerSet's local worker.
         """
-        if self.remote_workers():
-            weights = ray.put(self.local_worker().get_weights(policies))
-            for e in self.remote_workers():
-                e.set_weights.remote(weights)
+        if self.local_worker() is None and from_worker is None:
+            raise TypeError(
+                "No `local_worker` in WorkerSet, must provide `from_worker` "
+                "arg in `sync_weights()`!")
+
+        # Only sync if we have remote workers or `from_worker` is provided.
+        if self.remote_workers() or from_worker is not None:
+            weights = ray.put((from_worker
```

Comment:
saving from_worker in obj store? I thought we need weights here.

Reply:
This is correct (and the logic here wasn't touched in this PR):

1) ray.put places all the to-be-synched new weights in the object store and returns a object ref.
2) This object ref is used in the call to `set_weights.remote(...)`.
Note that if we had called `set_weights.remote()` with the actual weights (numpy), ray would have automatically transferred these to the object store first (and would have had to do so for all n remote workers, which is inefficient).

Reply:
oh I mis-read this. I thought we are putting from_worker in the obj store.
really we are just doing (from_worker or self.local_worker()) here. man I hate how yapf likes to break up long lines ...
sorry, and thanks.

Reply:
No, you were right! :) We should do both: Put it in the object store (`ray.put`) for better efficiency, but also keep the actual weights for synching to the local_worker, e.g. of the evaluation worker set!

Reply:
fixed